## **BONUS TASK**


This problem description is the same as Task 3. But there are some
technical challenges that you have to choose one.

1. Compare 3 different configurations while your model is
wider/deeper. Show and explain the performance result.
2. Compare 3 configurations for different Loss Function. Show and
explain your performance result.

3. Compare 3 configurations for the activation function. Show and
explain your performance result.

In [57]:
# Import Library yang dibutuhkan

import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [58]:
# Definisikan transformasi data
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# Muat dataset MNIST
train_dataset = datasets.MNIST(root='./data', train=True, transform=transform, download=True)
test_dataset = datasets.MNIST(root='./data', train=False, transform=transform, download=True)

# Muat data loader
batch_size = 64
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

In [59]:
# Fungsi untuk melatih model
def train_model(model, train_loader, criterion, optimizer, num_epochs):
    model.train()
    for epoch in range(num_epochs):
        running_loss = 0.0
        for inputs, labels in train_loader:
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()
            running_loss += loss.item()
        print(f'Epoch {epoch+1}/{num_epochs}, Loss: {running_loss/len(train_loader)}')

In [55]:
# Fungsi untuk mengevaluasi model
def evaluate_model(model, test_loader):
    model.eval()
    all_predictions = []
    all_labels = []

    with torch.no_grad():
        for images, labels in test_loader:
            outputs = model(images)
            _, predicted = torch.max(outputs, 1)

            all_predictions.extend(predicted.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())

    accuracy = accuracy_score(all_labels, all_predictions)
    confusion = confusion_matrix(all_labels, all_predictions)
    class_rep = classification_report(all_labels, all_predictions)

    return accuracy, confusion, class_rep

### ***1. Compare 3 different configurations while your model is wider/deeper. Show and explain the performance result.***

In [63]:
# Konfigurasi Model 1 (Lebar dan Dalam)
class Model1(nn.Module):
    def __init__(self):
        super(Model1, self).__init__()
        self.fc1 = nn.Linear(28*28, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 10)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = x.view(-1, 28*28)
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [64]:
# Konfigurasi Model 2 (Lebih Lebar)
class Model2(nn.Module):
    def __init__(self):
        super(Model2, self).__init__()
        self.fc1 = nn.Linear(28*28, 256)
        self.fc2 = nn.Linear(256, 128)
        self.fc3 = nn.Linear(128, 10)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = x.view(-1, 28*28)
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [65]:
# Konfigurasi Model 3 (Lebih Dalam)
class Model3(nn.Module):
    def __init__(self):
        super(Model3, self).__init__()
        self.fc1 = nn.Linear(28*28, 128)
        self.fc2 = nn.Linear(128, 128)
        self.fc3 = nn.Linear(128, 128)
        self.fc4 = nn.Linear(128, 10)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = x.view(-1, 28*28)
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.relu(self.fc3(x))
        x = self.fc4(x)
        return x

In [39]:
# Pelatihan dan Evaluasi Model 1
model1 = Model1()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model1.parameters(), lr=0.001)
num_epochs = 10
train_model(model1, train_loader, criterion, optimizer, num_epochs)
accuracy1, _, class_rep1 = evaluate_model(model1, test_loader)

Epoch 1/10, Loss: 0.40519687937679827
Epoch 2/10, Loss: 0.20259827782890436
Epoch 3/10, Loss: 0.14584674270970543
Epoch 4/10, Loss: 0.11981882021001884
Epoch 5/10, Loss: 0.09721967279970614
Epoch 6/10, Loss: 0.08480274702793658
Epoch 7/10, Loss: 0.07785453742145618
Epoch 8/10, Loss: 0.06922205952184796
Epoch 9/10, Loss: 0.06323463225941747
Epoch 10/10, Loss: 0.054849735868008516


In [40]:
# Pelatihan dan Evaluasi Model 2
model2 = Model2()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model2.parameters(), lr=0.001)
num_epochs = 10
train_model(model2, train_loader, criterion, optimizer, num_epochs)
accuracy2, _, class_rep2 = evaluate_model(model2, test_loader)

Epoch 1/10, Loss: 0.34442666483554507
Epoch 2/10, Loss: 0.15307433876409524
Epoch 3/10, Loss: 0.111466390741016
Epoch 4/10, Loss: 0.09126874967974656
Epoch 5/10, Loss: 0.07473855895468834
Epoch 6/10, Loss: 0.06633453107568553
Epoch 7/10, Loss: 0.0591479428954284
Epoch 8/10, Loss: 0.05392558632906514
Epoch 9/10, Loss: 0.04639040173325779
Epoch 10/10, Loss: 0.043696730310888665


In [41]:
# Pelatihan dan Evaluasi Model 3
model3 = Model3()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model3.parameters(), lr=0.001)
num_epochs = 10
train_model(model3, train_loader, criterion, optimizer, num_epochs)
accuracy3, _, class_rep3 = evaluate_model(model3, test_loader)

Epoch 1/10, Loss: 0.3745742074644832
Epoch 2/10, Loss: 0.17127199743840613
Epoch 3/10, Loss: 0.12842342012258037
Epoch 4/10, Loss: 0.10488218935092351
Epoch 5/10, Loss: 0.08865011812819204
Epoch 6/10, Loss: 0.07705588576330273
Epoch 7/10, Loss: 0.06791769566936796
Epoch 8/10, Loss: 0.06515849332761035
Epoch 9/10, Loss: 0.05761375319699683
Epoch 10/10, Loss: 0.052888454128793234


### EVALUASI MODEL

In [42]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Fungsi untuk mengevaluasi model
def evaluate_model(model, test_loader):
    model.eval()
    all_predictions = []
    all_labels = []

    with torch.no_grad():
        for images, labels in test_loader:
            outputs = model(images)
            _, predicted = torch.max(outputs, 1)

            all_predictions.extend(predicted.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())

    accuracy = accuracy_score(all_labels, all_predictions)
    confusion = confusion_matrix(all_labels, all_predictions)
    class_rep = classification_report(all_labels, all_predictions)

    return accuracy, confusion, class_rep

# Evaluasi Model 1
accuracy1, confusion1, class_rep1 = evaluate_model(model1, test_loader)

# Evaluasi Model 2
accuracy2, confusion2, class_rep2 = evaluate_model(model2, test_loader)

# Evaluasi Model 3
accuracy3, confusion3, class_rep3 = evaluate_model(model3, test_loader)

**MODEL 1**

In [43]:
# Cetak hasil evaluasi untuk masing-masing model
print("Model 1 Evaluation:")
print("\n")
print(f'Accuracy: {accuracy1:.4f}')
print('Confusion Matrix:')
print(confusion1)
print('Classification Report:')
print(class_rep1)

Model 1 Evaluation:


Accuracy: 0.9698
Confusion Matrix:
[[ 974    0    1    0    0    0    3    1    1    0]
 [   0 1122    1    4    0    1    0    4    1    2]
 [  12    2 1006    1    2    0    1    7    1    0]
 [   3    0    2  990    0    5    0    6    3    1]
 [   6    0    3    2  931    0    5    6    0   29]
 [   7    1    0   15    1  852    8    1    2    5]
 [  13    6    3    1    3    3  926    0    3    0]
 [   1    0    5    5    0    0    0 1008    1    8]
 [  14    0    5   18    6    7    6    5  904    9]
 [   6    1    0    6    4    3    0    4    0  985]]
Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.99      0.97       980
           1       0.99      0.99      0.99      1135
           2       0.98      0.97      0.98      1032
           3       0.95      0.98      0.96      1010
           4       0.98      0.95      0.97       982
           5       0.98      0.96      0.97       892
          

**MODEL 2**

In [44]:
print("Model 2 Evaluation:")
print("\n")
print(f'Accuracy: {accuracy2:.4f}')
print('Confusion Matrix:')
print(confusion2)
print('Classification Report:')
print(class_rep2)

Model 2 Evaluation:


Accuracy: 0.9756
Confusion Matrix:
[[ 973    0    1    0    0    0    1    1    3    1]
 [   0 1126    0    2    0    1    3    1    2    0]
 [   6    2  995    9    3    0    2    8    7    0]
 [   2    0    1  990    0    1    0    7    6    3]
 [   2    0    2    0  957    1    3    3    0   14]
 [   4    1    0   17    4  838    6    2   12    8]
 [   6    2    0    0    7    5  938    0    0    0]
 [   1    3    6    0    2    0    0 1000    1   15]
 [   7    0    1    2    3    1    2    5  947    6]
 [   2    1    0    3    4    2    1    4    0  992]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.99      0.98       980
           1       0.99      0.99      0.99      1135
           2       0.99      0.96      0.98      1032
           3       0.97      0.98      0.97      1010
           4       0.98      0.97      0.98       982
           5       0.99      0.94      0.96       892
          

**MODEL 3**

In [45]:
print("Model 3 Evaluation:")
print("\n")
print(f'Accuracy: {accuracy3:.4f}')
print('Confusion Matrix:')
print(confusion3)
print('Classification Report:')
print(class_rep3)

Model 3 Evaluation:


Accuracy: 0.9715
Confusion Matrix:
[[ 941    0    2    2    1   12    7    5    4    6]
 [   0 1126    0    2    0    2    1    1    3    0]
 [   1    2 1004    8    5    3    2    2    5    0]
 [   0    0    2  991    0    4    0    6    3    4]
 [   2    0    1    0  948    6    3    3    0   19]
 [   1    0    0   16    0  865    2    1    4    3]
 [   1    3    0    1    6   15  929    0    3    0]
 [   0    4    6   11    1    1    0 1001    0    4]
 [   1    0    1   14    0    5    1    3  944    5]
 [   1    2    0   12    5    7    0   10    6  966]]
Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.96      0.98       980
           1       0.99      0.99      0.99      1135
           2       0.99      0.97      0.98      1032
           3       0.94      0.98      0.96      1010
           4       0.98      0.97      0.97       982
           5       0.94      0.97      0.95       892
          

**PENJELASAN**
<br>

***Model 1 (Lebar dan Dalam)***
- Terdiri dari tiga layer linear (fully connected) dengan 128, 64, dan 10 neuron berturut-turut.
- Total layer 4 (termasuk input)
- Arsitektur: 28x28 (input) -> 128 -> 64 -> 10 (output).
- Model ini mencapai akurasi yang baik yaitu sekitar 96.98% pada dataset MNIST.

***Model 2 (Lebih lebar)***
- Terdiri dari tiga layer linear (fully connected) dengan 256, 128, dan 10 neuron berturut-turut.
- Total layer 4 (termasuk input)
- Arsitektur: 28x28 (input) -> 256 -> 128 -> 10 (output).
- Model ini mencapai akurasi yang lebih tinggi yaitu sekitar 97.56% dibandingkan dg Model 1.

***Model 3 (Lebih dalam)***
- Terdiri dari empat layer linear (fully connected) dengan 128 neuron pada setiap layer.
- Total layer 5 (termasuk input).
- Arsitektur: 28x28 (input) -> 128 -> 128 -> 128 -> 10 (output).
- Model ini mencapai akurasi yang lebih rendah dari Model ke 2 yaitu sekitar 97.15% tetapi memiliki akurasi lebih tinggi dari Model ke 1.


**KESIMPULAN**
- Model 2 (Lebih Lebar) memiliki potensi untuk menangani fitur lebih kompleks karena memiliki lebih banyak neuron pada layer pertama.
- Model 3 (Lebih Dalam) memiliki kedalaman yang lebih besar dengan lebih banyak layer yg dapat membantu dalam memahami representasi data yg lebih kompleks.
- Model 1 (Lebar dan Dalam) adalah model yang berada di tengah-tengah dalam hal kedalaman dan lebar.


Jadi, dari interpretasi di atas,  Model 2 adalah pilihan terbaik karena memiliki akurasi tertinggi dan kinerja yang baik secara keseluruhan dalam mengklasifikasikan data MNIST.


